### ISSUE TITLE
Notification: I am adding Acc into  haskell-perf/sequences banchmark comparison

### ISSUE THREAD

**Issue by Anton-Latukha** [nikita-volkov/acc#4]
Notifying so you can check that Acc is represented properly.

& so you are able to give commentary:

https://github.com/haskell-perf/sequences/pull/23

**Comment by Anton-Latukha** [comment]
After the merge & benchmark report update - it is good to make the announcement in public forum that people can now see real numbers on DList & Acc performance in operations. And bold claims do relatively add up. In my results there are: traverse claim doesn't hold water (if to believe `length` benchmark as traverse benchmark (https://github.com/haskell-perf/sequences/issues/24). `Acc` is somewhat faster than `Seq` in `cons&snoc` & >10x faster in `append` where it is on par even with `DList` `O(1)` `append`, overall very good results. `DList` is hard to overcome when `DList` has its laziness, when `DList` looses laziness - it becomes terrible, so `Acc` should hold the prime prize in the domain where operations/datatype are done strictly.

**Comment by nikita-volkov** [comment]
Thanks for the update. Overall a very insightful comparison! I'll give it a careful look some time soon. I won't be surprised if some optimisations come up later on since I haven't profiled the lib. This is the reason why I haven't made any announcements so far

**Comment by Anton-Latukha** [comment]
Nice to hear from you.
& the solution & design of the library I found great.

Then the report is solved for me.

**Comment by Anton-Latukha** [comment]
As knowing your position on things - I personally believe in `Acc` claim (but in a vague form) - as tests show - there is most probably a regression in the traverse or some difference, that is also why I notified authors of data types, so they check are my addition of benchmarks is done right.

For me is the most important thing for the public to know the true facts on data types now, then never. My agenda is to make Haskellers knowledgeable in which types apply where & make Haskellers stop hardcoding everything on `[]` & to migrate faster from it, which would make the products faster & code styles more flexible in the meantime. As in 2 GHC branches, there would be https://github.com/fumieval/Haskell-breaking-changes#upcoming-monomorphise-datalist - which is also a great moment for suiting data types to get a foothold in the ecosystem.

**Comment by nikita-volkov** [comment]
I've just updated the benchmark suite bundled with the package.
It compares against list, dlist and sequence by alternating between consing and snocing for construction and sum and length for reduction.

Following are the results:

```
sum/cons/1/acc                           mean 69.42 ns  ( +- 1.401 ns  )
sum/cons/1/list                          mean 68.33 ns  ( +- 2.317 ns  )
sum/cons/1/dlist                         mean 158.6 ns  ( +- 1.852 ns  )
sum/cons/1/sequence                      mean 73.25 ns  ( +- 1.562 ns  )
sum/cons/100/acc                         mean 2.230 μs  ( +- 47.92 ns  )
sum/cons/100/list                        mean 2.328 μs  ( +- 73.37 ns  )
sum/cons/100/dlist                       mean 4.611 μs  ( +- 65.33 ns  )
sum/cons/100/sequence                    mean 3.419 μs  ( +- 76.96 ns  )
sum/cons/10000/acc                       mean 409.3 μs  ( +- 6.000 μs  )
sum/cons/10000/list                      mean 644.7 μs  ( +- 144.7 μs  )
sum/cons/10000/dlist                     mean 1.097 ms  ( +- 17.30 μs  )
sum/cons/10000/sequence                  mean 916.3 μs  ( +- 131.3 μs  )
sum/snoc/1/acc                           mean 71.07 ns  ( +- 875.7 ps  )
sum/snoc/1/dlist                         mean 145.3 ns  ( +- 1.045 ns  )
sum/snoc/1/sequence                      mean 72.81 ns  ( +- 546.0 ps  )
sum/snoc/100/acc                         mean 2.953 μs  ( +- 86.24 ns  )
sum/snoc/100/dlist                       mean 4.019 μs  ( +- 26.64 ns  )
sum/snoc/100/sequence                    mean 3.335 μs  ( +- 58.80 ns  )
sum/snoc/10000/acc                       mean 599.6 μs  ( +- 5.346 μs  )
sum/snoc/10000/dlist                     mean 841.2 μs  ( +- 5.828 μs  )
sum/snoc/10000/sequence                  mean 865.4 μs  ( +- 10.13 μs  )
length/cons/1/acc                        mean 70.70 ns  ( +- 9.868 ns  )
length/cons/1/list                       mean 50.96 ns  ( +- 355.5 ps  )
length/cons/1/dlist                      mean 104.0 ns  ( +- 808.3 ps  )
length/cons/1/sequence                   mean 50.18 ns  ( +- 1.323 ns  )
length/cons/100/acc                      mean 2.111 μs  ( +- 63.50 ns  )
length/cons/100/list                     mean 958.6 ns  ( +- 5.650 ns  )
length/cons/100/dlist                    mean 3.088 μs  ( +- 42.06 ns  )
length/cons/100/sequence                 mean 2.434 μs  ( +- 53.62 ns  )
length/cons/10000/acc                    mean 397.4 μs  ( +- 3.094 μs  )
length/cons/10000/list                   mean 130.2 μs  ( +- 766.1 ns  )
length/cons/10000/dlist                  mean 550.0 μs  ( +- 3.412 μs  )
length/cons/10000/sequence               mean 770.9 μs  ( +- 11.94 μs  )
```

As you can see, in none of those does acc suffer from regression. Or am I missing your point?

Seems like further investigation is needed before any public announcements.

Oh! It is also worth mentioning that all benchmarks imply strict reduction. I.e., the container is evaluated to WHNF on each step.

**Comment by nikita-volkov** [comment]
The update suite is on the ["dev"](https://github.com/nikita-volkov/acc/tree/dev) branch.

**Comment by Anton-Latukha** [comment]
> the container is evaluated to WHNF on each step

https://github.com/haskell-perf/sequences/blob/2c8df3e824ed048c45b1732b63180580f924d708/Time.hs#L186 seems also do `whnf`. *idk* maybe a difference in that `foldl'` is applied. 
Maybe it is because you do: https://github.com/nikita-volkov/acc/blob/1a7304f01abc7befe2126f5775a1ec2c96bddc65/bench/Main.hs#L67 that `nf` & `haskell-perf` may not - it is on the fringe of my current knowledge, especially in the benchmarking, I just was doing work & had questions of the ballpark of data types performance & was curious & went into default place I know (`haskell-perf/sequence`) & did benchmarks in a uniform way of how they are done inside there & then send them the code. Learning both `criterion` & `gauge` at the same time is a bit to much for me currently until I learn at least one of them solidly.

But I know that much - that if enable `force` for `DList` in `Append` test - its metrics become terrible, which is expected.

But you got the idea. It is probably a nice thing to discuss/contribute to `haskell-perf/sequence` in any case, as there should be more different kinds of benchmarks on data types there. And as people would mostly navigate through that repo. I think that benchmarks both in the lazy end form (for lazy continuations) & in normalized end form (end-end form for processing every element) apply there.

There is no reason to hide projects you want to share with the community, if benchmarks are done properly & in proper detail - the truth would be revealed & stream would stear into the right direction. & currently I feel we starting to talk about who's benchmarks are more true approach, while I think if there are no errors - all benchmark modes apply.

As we now may probably talk about the `NFData` normalization benchmark.

**Comment by Anton-Latukha** [comment]
If you would like - you can just request a full revert of the [f41659b1f2e3e3a0a6a1ed61577d5a6122d2beed](https://github.com/haskell-perf/sequences/commit/f41659b1f2e3e3a0a6a1ed61577d5a6122d2beed) commit & regeneration of a result - that is what source version control is created for & what [Founding Wiki principles](https://meta.wikimedia.org/wiki/Founding_principles) show as productive, which I encourage - when my statements were found faulty (due to Gödel incompleteness proof - real-life statements would always be nontotal) - let other people make a proper revision.

**Comment by nikita-volkov** [comment]
Yeah. I agree that it's best to have the comparison benchmarks in a centralised place and haskell-perf is a great one. I've only placed the benchmarks here for now since it's easier to explore in your own package and to have a target for optimisation at hand.

Regarding

https://github.com/nikita-volkov/acc/blob/3ef7f98615d07e24f2846680d95df9c6428ca689/bench/Main.hs#L53-L69

Having that bit isolated into a non-inlinable function ensures that reducer and constructor implementations don't get fused.

**Comment by Anton-Latukha** [comment]
> reducer and constructor implementations don't get fused.

I understand why that is done.

`haskell-perf` does:
```haskell
sampleList i = evaluate $ force [1 .. i]
```
for all generators, which I guess also insures the generator not fusing into construction.

---

The more I think about it - the more it becomes apparent that datatypes deserve a lot of typization (typing) into different benchmarks. There are a lot of benchmarks that are not presented.

**Comment by Anton-Latukha** [comment]
Also we see here - that your suite benchmarks both construction & operation at the same time, `haskell-perf` may prebuild data & then benchmarks the operation. Again - construction is another thing to benchmark.

**Comment by nikita-volkov** [comment]
You're correct. I introduce the assumption that these data-structures will be used as intermediate builders intended for later reduction, which is the ultimate purpose of the "acc" package. I.e., a two-step process. Also I assume that these builders will be strictly evaluated during construction.

With that context it doesn't make sense to measure the steps in isolation, because in the end it is both steps that will affect the performance of your application. So isolated performance of each step is not what the user should be interested in (assuming the context). In case of the haskell-perf benchmarks, these steps are measured in isolation with some important aspects getting lost in the scope.

The primary reason of the degradation of the reduction benchmarks (length, max, min) of Acc and DList in haskell-perf seems to be the way their initial values get constructed prior to the execution of the benchmarks. I mean this:

https://github.com/haskell-perf/sequences/blob/2c8df3e824ed048c45b1732b63180580f924d708/Time.hs#L336-L337

In case of Acc, constructing from list balances the tree in such a way that the leftmost item is nested the deepest. The reduction algorithms OTOH are optimised for the opposite case. So the reduction involves complete rebalancing of the tree. I suspect similar reasons apply to DList as well. In real life that likely wouldn't be a problem because you wouldn't be constructing the intermediate data structure from list only to fold afterwards, you'd be directly folding the list in such case.

I also suspect that in case of Seq, constructing directly from list produces a Seq in the most balanced way, but since that process is excluded from the benchmark, we get the bias.

A possibly unrelated issue seems to be lurking from the use of `whnf` in some benchmarks:

https://github.com/haskell-perf/sequences/blob/2c8df3e824ed048c45b1732b63180580f924d708/Time.hs#L183-L203

So it seems like the problem is that the "haskell-perf" benchmarks introduce a random bias. And the reason for that is because they are not focused on a particular use-case (unlike the ones we have here). As the result we get a messy mix of lazy and strict evaluation and coverage of completely unrelated tasks like indexing and appending. I think the proper course of action is to reconsider the structure of the "haskell-perf" benchmarks, because currently they are likely to lead to wrong conclusions.

**Comment by nikita-volkov** [comment]
This whole issue has lead me to the following idea in regards to "acc": https://github.com/nikita-volkov/acc/issues/6. Happy to discuss it in that dedicated thread.

**Comment by Anton-Latukha** [comment]
Yes. That is what I say. My idea for contributors is to atomize & typize the `haskell-perf` testing to introduce atomic operation testing first. It reality - yes, there can be no ["One Size Fits All"](https://en.wikipedia.org/wiki/One_Size_Fits_All_(Frank_Zappa_album)) solution to data type testing in Haskell. `haskell-perf` code of course not suited from the get-go to consider all kinds of tests (I just tried to put `fromList` test there and failed so far). But `chrisdone` is openly acceptive of benchmarks & this is what `haskell-perf` is for in the first place - so they knew what they get into 8)

If you feel uneasy with the messy tests there - I would pull `Acc` from there myself, but that would be the simple but counterproductive step to expanding the test suite there to show where `Acc` actually works.

If `Acc` wants to be advertised properly - the documentation should state in precise terms (meaning more precise) in which context the x2 performance is guaranteed to work.

I'd liked for everyone to get along & expand the central public benchmarking suite for data types - because we anyway do that themselves on a per-package basis.

But currently I feel obligated to revert my submission of the data type there, which also would return `Acc` into being not discoverable compared to other data types represented there (as I find myself frequently going into `haskell-perf` projects to remind myself things.

**Comment by Anton-Latukha** [comment]
Yep. I think this thread can be closed.

**Comment by nikita-volkov** [comment]
I actually don't think that Acc should be removed from haskell-perf. Instead we better find a way of correcting the benchmarks there so that they convey correct information.

E.g., if the [`sampleAcc`](https://github.com/haskell-perf/sequences/blob/2c8df3e824ed048c45b1732b63180580f924d708/Time.hs#L336-L337) function is reimplemented to construct Acc by consing, you'll get it balanced perfectly and the reduction will perform at its best (which is what we get for Seq there). Of course, that wouldn't make those benchmarks more informative, but at least that would remove the negative bias. Same could be done in regards to DList. That could be the first step. Ultimately though, again, I believe that the whole suite needs to be redesigned. That should be the next step IMO.

**Comment by nikita-volkov** [comment]
I've realised that since the purpose of Acc is for aggregating data for reduction, and since reduction is almost always from left, it is a good default to have the fromList conversion doing the heavy lifting of rebalancing. Which essentially means that the fromList operation needs to reverse the list internally.

I've reimplemented that part and made a release 0.1.3.1 with the updated fromList. Now simply rerunning the haskell-perf benchmarks with the latest version of acc should produce nice results.

@Anton-Latukha Can you please PR with according updates to haskell-perf and its results?

**Comment by nikita-volkov** [comment]
NB I've just released version 0.2, which removes the Generic instances and makes Acc non-strict about the element, preserving the strictness of the spine. The latter thing may affect the benchmarks as well.

**Comment by nikita-volkov** [comment]
I've extended the benchmark suite here and added a detailed analysis to [the readme](https://github.com/nikita-volkov/acc/blob/3401cb21e2aac0e83a9fd26ae30d706a8c89264d/README.md). @Anton-Latukha please take a look.

I think it's worth migrating this suite as a dedicated repo to "haskell-perf" and redirecting to it from the "sequences" repo. I'll bring it up for discussion there.
